src/Exception/BatchException.php

```diff
@@ -2,61 +2,136 @@
+
 /**
  * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
  */
-final class BatchException extends TransferException
+final class BatchException extends RuntimeException
 {
     /**
-     * @var TransferException[]
+     * @var BatchResult
+     */
+    private $result;
+
+    /**
+     * @var \SplObjectStorage
      */
     private $exceptions;
 
+    public function __construct()
+    {
+        $this->exceptions = new \SplObjectStorage();
+    }
+
     /**
-     * @param TransferException[] $exceptions
+     * Returns a list BatchResult which contains succesful responses
+     *
+     * @return BatchResult
      */
-    public function __construct(array $exceptions = [])
+    public function getResult()
     {
-        parent::__construct('An error occurred when sending multiple requests.');
+        return $this->result;
+    }
 
-        foreach ($exceptions as $e) {
-            if (!$e instanceof TransferException) {
-                throw new InvalidArgumentException('Exception is not an instanceof Http\Client\Exception\TransferException');
-            }
-        }
+    /**
+     * Sets the successful response list
+     *
+     * @param BatchResult $result
+     *
+     * @internal
+     */
+    public function setResult(BatchResult $result)
+    {
+        $this->result = $result;
+    }
+
+    /**
+     * Checks if a request is successful
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function isSuccessful(RequestInterface $request)
+    {
+        return isset($this->result) && $this->result->hasResponseFor($request);
+    }
 
-        $this->exceptions = $exceptions;
+    /**
+     * Checks if a request is failed
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function isFailed(RequestInterface $request)
+    {
+        return $this->exceptions->contains($request);
     }
 
     /**
      * Returns all exceptions
      *
-     * @return TransferException[]
+     * @return Exception[]
      */
     public function getExceptions()
     {
-        return $this->exceptions;
+        $exceptions = [];
+
+        foreach ($this->exceptions as $request) {
+            $exceptions[] = $this->exceptions[$request];
+        }
+
+        return $exceptions;
     }
 
     /**
-     * Checks if a specific exception exists
+     * Returns an exception for a request or null if not found
      *
-     * @param TransferException $exception
+     * @param RequestInterface $request
      *
-     * @return boolean TRUE if there is the exception else FALSE.
+     * @return Exception|null
      */
-    public function hasException(TransferException $exception)
+    public function getExceptionFor(RequestInterface $request)
     {
-        return array_search($exception, $this->exceptions, true) !== false;
+        if ($this->exceptions->contains($request)) {
+            return $this->exceptions[$request];
+        }
```

Comment:
returning null is a horrible thing imho :-)
as it makes errors go unnoticed. i rather have an exception with a
stacktrace than "fatal error: called method on something that is not an
object"